fedimint-core/src/module/mod.rs

```diff
@@ -618,10 +619,28 @@ impl<CI> ConsensusProposal<CI> {
     }
 }
 
+/// Module associated types required by both client and server
+pub trait ModuleTypes {
+    type Input: Input;
+    type Output: Output;
+    type OutputOutcome: OutputOutcome;
+    type ConsensusItem: ModuleConsensusItem;
+
+    fn decoder_builder() -> DecoderBuilder {
+        let mut decoder_builder = Decoder::builder();
+        decoder_builder.with_decodable_type::<Self::Input>();
+        decoder_builder.with_decodable_type::<Self::Output>();
+        decoder_builder.with_decodable_type::<Self::OutputOutcome>();
+        decoder_builder.with_decodable_type::<Self::ConsensusItem>();
+        decoder_builder
+    }
+}
+
 #[apply(async_trait_maybe_send!)]
 pub trait ServerModule: Debug + Sized {
+    type ModuleTypes: ModuleTypes;
```

Comment:
What's the root reason why we need another level of indirection instead of inline? because it is shared between client and server? In that case maybe this should be `Common: ModuleCommon`?

Reply:
Yes, pretty much. I want to get rid of the client module having an associated type which is the server module type. Naming can be changed.

Reply:
Renamed.